---
output: github_document
---

<!-- README.md is generated from README.Rmd. Please edit that file -->

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.path = "man/figures/README-",
  out.width = "100%"
)
```

# fetchNOMe

<!-- badges: start -->
[![R-CMD-check](https://github.com/fmi-basel/gpeters-fetchNOMe/actions/workflows/R-CMD-check.yaml/badge.svg)](https://github.com/fmi-basel/gpeters-fetchNOMe/actions/workflows/R-CMD-check.yaml)
<!-- badges: end -->

fetchNOMe R package provides functionality for fast retrieval of NOMe-seq data (M.CviPI-induced GpC methylation as well as endogenous CpG methylation) from BAM files generated by pipelines for alignment of bisulfite converted cytosine methylation data.

## Installation

You can install the development version of fetchNOMe from [GitHub](https://github.com/) with:

``` r
# install.packages("devtools")
devtools::install_github("fmi-basel/gpeters-fetchNOMe")
```

## Example

Comes lates...

```{r example}
library(fetchNOMe)
## basic example code
```
