
#' Fetch protection aggregated statistics from NOMe-seq BAM files
#'
#' Function retrieves statistics of number of protected GCHs out of total number of GCHs in fragments in a set of BAM files generated for NOMe-seq data and set of regions of interest.
#' This statistics is mainly needed for analysis of NOMe-seq data generated for naked DNA controls, such as lambda DNA, and used for estimating noise in background and create background model for later use during footprint inference or footprint prediction.
#'
#' @param bamfiles \code{character} paths to bam files
#' @param samplenames \code{character} names of samples. If \code{collapseBySample} is \code{TRUE} counts are aggregated across bam files with same sample name.
#' @param regions \linkS4class{GRanges} object for regions of interest.
#' @param genome \code{character} path to fasta file which was used as reference for alignment.
#' @param alignerUsed \code{character} that defines which aligner was used to generate BAM files. Currently supported aligners are QuasR, Bismark and BISCUIT.
#' @param collapseBySample \code{logical} indicating whether to collapse counts for bam files with same \code{samplenames}. If \code{FALSE} prefix \code{s} followed by index is added to \code{samplenames}.
#' @param remove_nonunique \code{logical} if \code{TRUE} only unique fragments are analyzed. Uniqueness is defined by fragments start, end and methylation states of all cytosines. 
#' @param clip_until_nbg \code{integer} controlling clipping partial footprints at both ends of fragments. Namely, protected states are erased from each end until \code{clip_until_nbg} unprotected states are met.
#' Set \code{clip_until_nbg=0L} to skip clipping preprocessing.
#' @param noclip_protect_frac_above \code{numeric} controlling fragment clipping, i.e. fragments which have fraction of protected states higher than \code{max_protect_frac} escape clipping. This parameter allows to avoid removal of fully protected fragments during clipping process.
#' @param max_bisC_meth \code{numeric} controlling removal of fragments with failed bisulfite conversion, i.e. fragments with fraction of methylated non-GCH, and non-WCG cytosines higher then \code{max_bisC_meth} are removed from the analysis to get rid of fragments for which bisulfite conversion failed.
#' @param min_bisC_size \code{integer} setting minimum number non-GCH, and non-WCG cytosines for filtering according to bisulfite conversion efficiency, 
#' i.e. only fragments with number of non-GCH, and non-WCG cytosines higher then \code{min_bisC_size} are subject to filtering based on bisulfite conversion efficiency.
#' @param mapqMin \code{integer} setting minimum MAPQ of reads to be included into analysis.
#' @param mapqMax \code{integer} setting maximum MAPQ of reads to be included into analysis.
#' @param max_read_size maximum read length expected in the BAM file. This parameter only controls how much regions of interest are flanked left and right to extract reference sequences.
#' This parameter does not influence any filtering of fragments based on the length.
#' @param ncores number of cores to use.
#'
#' @return \code{tibble} where each row corresponds to sample - region combination. 
#' Columns of the returned \code{tibble} represent:
#' \describe{
#'   \item{SampleName}{Name of a sample as defined by \code{samplenames}}.
#'   \item{seqnames, start, end}{seqnames (or chromosomes) and coordinates of corresponding regions.}
#'   \item{names}{the same as \code{names(regions)} if it is not \code{NULL} or index of a corresponding region in \code{regions}.}
#'   \item{nFragsFetched}{number of fetched fragments before filtering.}
#'   \item{nFragsNonUnique}{number of non-unique fragments as defined by coordinates and methylation profiles.}
#'   \item{nFragsBisFailed}{number of fragments for which bisulfite conversion may have failed, namely non-GCH and non-WCG methylation level is higher than \code{max_bisC_meth}.}
#'   \item{nFragsAnalyzed}{number of fragments analysed for co-occurrence table after filtering non-unique and bisulfite conversion failed fragments.}
#'   \item{ClippedUntilNbg}{parameter \code{clip_until_nbg} used for clipping partial footprints from both ends.}
#'   \item{NotClippedProtectAbove}{parameter \code{noclip_protect_frac_above} used for clipping partial footprints from both ends.}
#'   \item{nFragsClipped}{number of fragments clipped from both ends to get rid of paetial footprints.}
#'   \item{ProtectStats}{\code{matrix} with columns reflecting combinations of total number and number of protected GCHs:
#'   \describe{
#'   \item{TotalGCH}{total number of GCHs.}
#'   \item{ProtectedGCH}{number of of protected GCHs.}
#'   \item{Nfrags}{number of fragments met with combination of \code{TotalGCH} and \code{ProtectedGCH}.}
#'   }
#'   }
#' }

#' @export
#'
get_protect_stats_from_bams <- function(bamfiles,
																							samplenames,
																							regions,
																							genome,
																							alignerUsed = c("QuasR","Bismark","BISCUIT"),
																							collapseBySample = TRUE,
																							remove_nonunique = TRUE,
																							clip_until_nbg  = 1L,
																							noclip_protect_frac_above = 0.90,
																							max_bisC_meth = 0.1,
																							min_bisC_size = 10,
																							mapqMin = 0,
																							mapqMax = 255,
																							max_read_size = 1000L,
																							ncores = 1L){
	if(!inherits(bamfiles,"character"))
		stop("bamfiles must be vector of character")
	if(!inherits(samplenames,"character"))
		stop("samplenames must be vector of character")
	if(any(!file.exists(bamfiles))){
		nonex <- which(!file.exists(bamfiles))
		nonex <- paste0(bamfiles[nonex],collapse = "\n")
		stop(paste0("Could not find following BAM files:\n",nonex))
	}
	
	if(!inherits(regions,"GRanges"))
		stop("regions must be Granges")
	
	if(!inherits(genome,"character") & length(genome) !=1)
		stop("genome must be character and contain only one element")
	
	if(!file.exists(genome))
		stop("Could not find genome:", genome)
	
  alignerUsed <- match.arg(alignerUsed)
  
  message(paste0("Collecting protection statistics from BAM files generated by ",alignerUsed,"."))
  
  
	## load refsequences for regions
	## extend regions by max_read_size from left and right
	
	fa_index <- Rsamtools::scanFaIndex(genome)
	GenomeInfoDb::seqlevels(regions) <- GenomeInfoDb::seqlevels(fa_index)
	GenomeInfoDb::seqinfo(regions) <- suppressMessages(GenomeInfoDb::Seqinfo(seqnames = as.character(GenomeInfoDb::seqnames(fa_index)),
																														 seqlengths = GenomicRanges::width(fa_index)))
	refseq_gr <- suppressWarnings(GenomicRanges::resize(regions,width = GenomicRanges::width(regions) + 2*max_read_size,fix="center"))
	
	## trim to remove parts out of chromosomes
	refseq_gr <- GenomicRanges::trim(refseq_gr)
	refseqs <- Rsamtools::scanFa(genome,refseq_gr)
	
	# 2. split vector by sample name. If collapseBySample is FALSE add to samplenames some suffixes
	if(!collapseBySample){
		samplenames <- paste0("s",seq_along(samplenames),"_",samplenames)
	}
	
	bamfiles <- split(bamfiles, f=samplenames)
	# 3. for each set of bam files call C++ function
	dataMatr_list <- do.call(rbind,lapply(names(bamfiles),
																				function(bamgroupname){
																					ctbl <- do.call(rbind,parallel::mclapply(seq_along(regions),
																																				 function(regi){
																																				 	
																																				 	regdf <- GenomicRanges::as.data.frame(regions[regi])
																																				 	regdf$seqnames <- as.character(regdf[,"seqnames"])
																																				 	seqgr <- as.data.frame(refseq_gr[regi])
																																				 	
																																				 	outlist <- fetch_protect_stats_from_bams_cpp(infiles = bamfiles[[bamgroupname]],
																																				 																							 regionChr = regdf[1,"seqnames"],
																																				 																							 regionStart = regdf[1,"start"] - 1, # convert to 0-based
																																				 																							 regionEnd = regdf[1,"end"],
																																				 																							 seqstring = as.character(refseqs[regi]),
																																				 																							 seqStart = seqgr[1,"start"] - 1,
																																				 																							 seqEnd = seqgr[1,"end"],
																																				 																							 remove_nonunique = remove_nonunique,
																																				 																							 clip_until_nbg  = clip_until_nbg,
																																				 																							 max_protect_frac = noclip_protect_frac_above,
																																				 																							 max_bisC_meth = max_bisC_meth,
																																				 																							 min_bisC_size = min_bisC_size,
																																				 																							 mapqMin = mapqMin,
																																				 																							 mapqMax = mapqMax,
																																				 																							 alignerUsed = alignerUsed)
																																				 	
																																				 	data_mat <- outlist[["ProtectStats"]]
																																				 	
																																				 	outtbl <- tibble::tibble_row("SampleName" = bamgroupname,
																																				 															 "seqnames" = regdf[1,"seqnames"],
																																				 															 "start" = regdf[1,"start"],
																																				 															 "end" = regdf[1,"end"],
																																				 															 "strand" = regdf[1,"strand"],
																																				 															 "names" = ifelse(!is.null(names(regions)), names(regions)[regi],regi),
																																				 															 "nFragsFetched" = outlist[["nFragsFetched"]],
																																				 															 "nFragsNonUnique" = outlist[["nFragsNonUnique"]],
																																				 															 "nFragsBisFailed" = outlist[["nFragsBisFailed"]],
																																				 															 "nFragsAnalyzed" = outlist[["nFragsAnalyzed"]],
																																				 															 "ClippedUntilNbg" = outlist[["ClippedUntilNbg"]],
																																				 															 "NotClippedProtectAbove" = outlist[["NotClippedProtectAbove"]],
																																				 															 "nFragsClipped" = outlist[["nFragsClipped"]],
																																				 															 "ProtectStats" = list(data_mat))
																																				 	
																																				 	return(outtbl)
																																				 },mc.cores = ncores))
																					
																					return(ctbl)
																				}))
}